server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSReadqOptionsCheckUtility.java

```diff
@@ -60,36 +58,41 @@ public CICSReadqOptionsCheckUtility(
      * @param <E> A subclass of ParserRuleContext
      */
     public <E extends ParserRuleContext> void checkOptions(E ctx) {
-        if (ctx.getRuleIndex() == RULE_cics_readq_td)
-            checkTd((CICSParser.Cics_readq_tdContext) ctx);
-        else if (ctx.getRuleIndex() == RULE_cics_readq_ts)
-            checkTs((CICSParser.Cics_readq_tsContext) ctx);
+        if (ctx.getRuleIndex() == RULE_cics_readq_ts_td) {
+            checkOpts((CICSParser.Cics_readq_ts_tdContext) ctx);
+        }
 
         checkDuplicates(ctx);
     }
 
-    private void checkTd(CICSParser.Cics_readq_tdContext ctx) {
-        if (!ctx.TD().isEmpty())
-            checkHasMandatoryOptions(ctx.QUEUE(), ctx, "QUEUE");
-        if (!ctx.NOSUSPEND().isEmpty())
-            checkHasMandatoryOptions(ctx.TD(), ctx, "TD");
-        if (!ctx.cics_into_set().isEmpty())
-            checkSetTd(ctx.cics_into_set().listIterator().next());
-
+    private void checkOpts(CICSParser.Cics_readq_ts_tdContext ctx) {
+        if (!ctx.TD().isEmpty()) {
+            checkTd(ctx);
+        } else {
+            checkTs(ctx);
+        }
+    }
+    private void checkTd(CICSParser.Cics_readq_ts_tdContext ctx) {
+        checkHasMandatoryOptions(ctx.QUEUE(), ctx, "QUEUE");
         checkHasExactlyOneOption("INTO or SET", ctx, ctx.cics_into_set());
+        checkHasIllegalOptions(ctx.ITEM(), "ITEM");
+        checkHasIllegalOptions(ctx.NEXT(), "NEXT");
+        checkHasIllegalOptions(ctx.NUMITEMS(), "NUMITEMS");
+        checkHasIllegalOptions(ctx.QNAME(), "QNAME");
+        checkHasIllegalOptions(ctx.TS(), "TS");
```

Comment:
I think that clearer messages would be generated if the `TS`/`TD` exclusivity check was done already in the `checkOpts` (perhaps via `checkHasMutuallyExclusiveOptions`).

Reply:
would this be clearer ? 42c216a8b3b6f30581811cc244e51d09477d9f79